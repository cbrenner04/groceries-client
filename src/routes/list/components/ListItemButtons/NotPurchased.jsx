import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { listTypeToSnakeCase } from '../../../../utils/format';
import { Bookmark, Complete, Edit, Trash } from '../../../../components/ActionButtons';

function NotPurchasedItemButtons(props) {
  const listItemPath = () => {
    const listId = props.item[`${listTypeToSnakeCase(props.listType)}_id`];
    return `/lists/${listId}/${listTypeToSnakeCase(props.listType)}_items`;
  };

  const handleRead = () => props.handleReadOfItem(props.item);
  const handleUnRead = () => props.handleUnReadOfItem(props.item);

  return (
    <ButtonGroup className="float-right">
      {props.listType === 'BookList' && (
        <Bookmark handleClick={props.item.read ? handleUnRead : handleRead} read={props.item.read} />
      )}
      <Complete handleClick={() => props.handlePurchaseOfItem(props.item)} />
      <Edit to={`${listItemPath()}/${props.item.id}/edit`} />
      <Trash handleClick={() => props.handleItemDelete(props.item)} />
    </ButtonGroup>
  );
}

NotPurchasedItemButtons.propTypes = {
  listType: PropTypes.string.isRequired,
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    read: PropTypes.bool,
  }).isRequired,
  handlePurchaseOfItem: PropTypes.func.isRequired,
  handleItemDelete: PropTypes.func.isRequired,
  handleReadOfItem: PropTypes.func.isRequired,
  handleUnReadOfItem: PropTypes.func.isRequired,
};

export default NotPurchasedItemButtons;
