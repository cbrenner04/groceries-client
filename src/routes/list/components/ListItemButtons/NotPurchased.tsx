import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Bookmark, Complete, EditButton, Trash } from '../../../../components/ActionButtons';
import { listItem } from '../../../../prop-types';

function NotPurchasedItemButtons(props) {
  return (
    <ButtonGroup className={`${props.multiSelect ? 'list-item-buttons' : ''} float-end`}>
      {props.listType === 'BookList' && (
        <Bookmark
          handleClick={() => props.toggleItemRead(props.item)}
          read={props.item.read}
          testID={`not-purchased-item-${props.item.read ? 'unread' : 'read'}-${props.item.id}`}
          disabled={props.pending}
        />
      )}
      <Complete
        handleClick={() => props.handlePurchaseOfItem(props.item)}
        testID={`not-purchased-item-complete-${props.item.id}`}
        disabled={props.pending}
      />
      <EditButton
        handleClick={() => props.handleItemEdit(props.item)}
        testID={`not-purchased-item-edit-${props.item.id}`}
        disabled={props.pending}
      />
      <Trash
        handleClick={() => props.handleItemDelete(props.item)}
        testID={`not-purchased-item-delete-${props.item.id}`}
        disabled={props.pending}
      />
    </ButtonGroup>
  );
}

NotPurchasedItemButtons.propTypes = {
  listType: PropTypes.string.isRequired,
  item: listItem.isRequired,
  handlePurchaseOfItem: PropTypes.func.isRequired,
  handleItemDelete: PropTypes.func.isRequired,
  toggleItemRead: PropTypes.func.isRequired,
  handleItemEdit: PropTypes.func.isRequired,
  pending: PropTypes.bool.isRequired,
  multiSelect: PropTypes.bool,
};

export default NotPurchasedItemButtons;
