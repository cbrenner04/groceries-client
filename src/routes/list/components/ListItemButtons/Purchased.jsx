import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Bookmark, Refresh, EditButton, Trash } from '../../../../components/ActionButtons';
import { listItem } from '../../../../types';

const PurchasedItemButtons = (props) => (
  <ButtonGroup className={`${props.multiSelect ? 'list-item-buttons' : ''} float-end`}>
    {['GroceryList', 'SimpleList', 'ToDoList'].includes(props.listType) && (
      <Refresh
        handleClick={() => props.handleItemRefresh(props.item)}
        testID={`purchased-item-refresh-${props.item.id}`}
        disabled={props.pending}
      />
    )}
    {props.listType === 'BookList' && (
      <Bookmark
        handleClick={() => props.toggleItemRead(props.item)}
        read={props.item.read}
        testID={`purchased-item-${props.item.read ? 'unread' : 'read'}-${props.item.id}`}
        disabled={props.pending}
      />
    )}
    <EditButton
      handleClick={() => props.handleItemEdit(props.item)}
      testID={`purchased-item-edit-${props.item.id}`}
      disabled={!props.multiSelect || props.selectedItems.length === 0 || props.pending}
    />
    <Trash
      handleClick={() => props.handleItemDelete(props.item)}
      testID={`purchased-item-delete-${props.item.id}`}
      disabled={props.pending}
    />
  </ButtonGroup>
);

PurchasedItemButtons.propTypes = {
  listType: PropTypes.string.isRequired,
  item: listItem.isRequired,
  handleItemRefresh: PropTypes.func.isRequired,
  handleItemDelete: PropTypes.func.isRequired,
  toggleItemRead: PropTypes.func.isRequired,
  handleItemEdit: PropTypes.func.isRequired,
  multiSelect: PropTypes.bool.isRequired,
  selectedItems: PropTypes.arrayOf(listItem).isRequired,
  pending: PropTypes.bool.isRequired,
};

export default PurchasedItemButtons;
