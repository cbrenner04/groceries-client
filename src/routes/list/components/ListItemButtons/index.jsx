import React from 'react';
import PropTypes from 'prop-types';

import NotPurchased from './NotPurchased';
import Purchased from './Purchased';

const ListItemButtons = (props) =>
  props.purchased ? (
    <Purchased
      listType={props.listType}
      item={props.item}
      handleItemUnPurchase={props.handleItemUnPurchase}
      handleItemDelete={props.handleItemDelete}
      handleReadOfItem={props.handleReadOfItem}
      handleUnReadOfItem={props.handleUnReadOfItem}
    />
  ) : (
    <NotPurchased
      listType={props.listType}
      item={props.item}
      handlePurchaseOfItem={props.handlePurchaseOfItem}
      handleItemDelete={props.handleItemDelete}
      handleReadOfItem={props.handleReadOfItem}
      handleUnReadOfItem={props.handleUnReadOfItem}
    />
  );

ListItemButtons.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    read: PropTypes.bool,
  }).isRequired,
  purchased: PropTypes.bool,
  handleItemDelete: PropTypes.func.isRequired,
  handlePurchaseOfItem: PropTypes.func.isRequired,
  handleReadOfItem: PropTypes.func.isRequired,
  handleUnReadOfItem: PropTypes.func.isRequired,
  handleItemUnPurchase: PropTypes.func.isRequired,
  listType: PropTypes.string.isRequired,
};

export default ListItemButtons;
