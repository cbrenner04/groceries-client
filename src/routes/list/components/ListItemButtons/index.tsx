import React from 'react';

import NotPurchased from './NotPurchased';
import Purchased from './Purchased';
import type { IListItem, EListType } from '../../../../typings';

export interface IListItemButtonsProps {
  item: IListItem;
  purchased: boolean;
  handleItemDelete: (item: IListItem) => void;
  handlePurchaseOfItem: (item: IListItem) => void;
  handleItemRefresh: (item: IListItem) => void;
  listType: EListType;
  toggleItemRead: (item: IListItem) => void;
  handleItemEdit: (item: IListItem) => void;
  multiSelect: boolean;
  selectedItems: IListItem[];
  pending: boolean;
}

const ListItemButtons: React.FC<IListItemButtonsProps> = (props): React.JSX.Element =>
  props.purchased ? (
    <Purchased
      listType={props.listType}
      item={props.item}
      handleItemRefresh={props.handleItemRefresh}
      handleItemDelete={props.handleItemDelete}
      toggleItemRead={props.toggleItemRead}
      handleItemEdit={props.handleItemEdit}
      multiSelect={props.multiSelect}
      selectedItems={props.selectedItems}
      pending={props.pending}
    />
  ) : (
    <NotPurchased
      listType={props.listType}
      item={props.item}
      handlePurchaseOfItem={props.handlePurchaseOfItem}
      handleItemDelete={props.handleItemDelete}
      toggleItemRead={props.toggleItemRead}
      handleItemEdit={props.handleItemEdit}
      pending={props.pending}
      multiSelect={props.multiSelect}
    />
  );

export default ListItemButtons;
