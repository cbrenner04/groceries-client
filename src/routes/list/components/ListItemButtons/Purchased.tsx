import React from 'react';
import { ButtonGroup } from 'react-bootstrap';

import { Bookmark, Refresh, EditButton, Trash } from '../../../../components/ActionButtons';
import { IListItem } from '../../../../typings';

interface IPurchasedItemButtonsProps {
  listType: string;
  item: IListItem;
  handleItemRefresh: (item: IListItem) => void;
  handleItemDelete: (item: IListItem) => void;
  toggleItemRead: (item: IListItem) => void;
  handleItemEdit: (item: IListItem) => void;
  multiSelect: boolean;
  selectedItems: IListItem[];
  pending: boolean;
}

const PurchasedItemButtons: React.FC<IPurchasedItemButtonsProps> = (props) => (
  <ButtonGroup className={`${props.multiSelect ? 'list-item-buttons' : ''} float-end`}>
    {['GroceryList', 'SimpleList', 'ToDoList'].includes(props.listType) && (
      <Refresh
        handleClick={() => props.handleItemRefresh(props.item)}
        testID={`purchased-item-refresh-${props.item.id}`}
        disabled={props.pending}
      />
    )}
    {props.listType === 'BookList' && (
      <Bookmark
        handleClick={() => props.toggleItemRead(props.item)}
        read={props.item.read ?? false}
        testID={`purchased-item-${props.item.read ? 'unread' : 'read'}-${props.item.id}`}
      />
    )}
    <EditButton
      handleClick={() => props.handleItemEdit(props.item)}
      testID={`purchased-item-edit-${props.item.id}`}
      disabled={!props.multiSelect || props.selectedItems.length === 0 || props.pending}
    />
    <Trash
      handleClick={() => props.handleItemDelete(props.item)}
      testID={`purchased-item-delete-${props.item.id}`}
      disabled={props.pending}
    />
  </ButtonGroup>
);

export default PurchasedItemButtons;
