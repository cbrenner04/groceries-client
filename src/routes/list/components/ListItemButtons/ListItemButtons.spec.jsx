import React from 'react';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import ListItemButtons from './index';

describe('ListItemButtons', () => {
  let props;

  const renderListItemButtons = (localProps) => {
    const history = createMemoryHistory();
    return render(
      <Router history={history}>
        <ListItemButtons {...localProps} />
      </Router>,
    );
  };

  beforeEach(() => {
    props = {
      item: {
        grocery_list_id: 1,
        id: 1,
        read: true,
      },
      purchased: true,
      handleItemDelete: jest.fn(),
      handlePurchaseOfItem: jest.fn(),
      toggleItemRead: jest.fn(),
      handleItemUnPurchase: jest.fn(),
      listType: 'GroceryList',
    };
  });

  it('renders Purchased when the item is purchased', () => {
    props.purchased = true;
    const { container, getAllByRole } = renderListItemButtons(props);
    const buttons = getAllByRole('button');

    expect(container).toMatchSnapshot();
    expect(buttons.length).toBe(2);
    expect(buttons[0].firstChild).toHaveClass('fa-redo');
    expect(buttons[1].firstChild).toHaveClass('fa-trash');
  });

  it('renders NotPurchased when the item is not purchased', () => {
    props.purchased = false;
    const { container, getAllByRole, getByRole } = renderListItemButtons(props);
    const buttons = getAllByRole('button');

    expect(container).toMatchSnapshot();
    expect(buttons.length).toBe(2);
    expect(buttons[0].firstChild).toHaveClass('fa-check');
    expect(buttons[1].firstChild).toHaveClass('fa-trash');
    expect(getAllByRole('link').length).toBe(1);
    expect(getByRole('link').firstChild).toHaveClass('fa-edit');
  });
});
