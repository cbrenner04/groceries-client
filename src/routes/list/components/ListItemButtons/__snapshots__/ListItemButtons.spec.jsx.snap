// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemButtons renders NotPurchased when the item is not purchased 1`] = `
<div>
  <div
    class=" float-end btn-group"
    role="group"
  >
    <button
      class="p-0 me-3 btn btn-link"
      data-test-id="not-purchased-item-complete-id1"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <button
      class="p-0 me-3 btn btn-link"
      data-test-id="not-purchased-item-edit-id1"
      type="button"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="not-purchased-item-delete-id1"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`ListItemButtons renders Purchased when the item is purchased 1`] = `
<div>
  <div
    class=" float-end btn-group"
    role="group"
  >
    <button
      class="p-0 me-3  btn btn-link"
      data-test-id="purchased-item-refresh-id1"
      type="button"
    >
      <i
        class="fa fa-redo fa-2x text-primary"
      />
    </button>
    <button
      class="p-0 me-3 btn btn-link"
      data-test-id="purchased-item-edit-id1"
      disabled=""
      type="button"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="purchased-item-delete-id1"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;
